File: ide-features-trainer/res/learnProjects/java/LearnProject/src/BufferedReader.java
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.io.InputStream;
import java.util.List;

class BufferedReader {
  private InputStream inputStream;
  private int maxBufferSize = 1024;
  private byte[] buffer;

  public BufferedReader(InputStream inputStream) {
    this.inputStream = inputStream;
  }

  public BufferedReader(InputStream inputStream, int maxBufferSize) {
    this.inputStream = inputStream;
    this.maxBufferSize = maxBufferSize;
  }

  /**
   * @return one byte from input stream
   */
  public byte read() {
    throw new NotImplementedException();
  }

  /**
   * @return n bytes from input stream
   */
  public byte[] read(int n) {
    throw new NotImplementedException();
  }

  /**
   * @return one line from input stream as String
   */
  public String readLine() {
    throw new NotImplementedException();
  }

  /**
   * @return all lines from input stream as List of Strings
   */
  public List<String> lines() {
    throw new NotImplementedException();
  }
}

File: ide-features-trainer/ruby-features-trainer/resources/learnProjects/ruby/RubyMineLearningProject/src/buffered_reader.rb

class BufferedReader
  def initialize(input_stream, max_buffer_size = 1024)
    @input_stream = input_stream
    @max_buffer_size = max_buffer_size
  end

  @buffer = []

  def read(n = 1)
    # returns n bytes from buffer or try read max_buffer_size bytes
    # from input stream and return n bytes
  end

  def read_line
    # returns one line from input stream as string
  end

  def lines
    # returns all lines from input stream as list of strings
  end
end